api/cpp/include/slint.h

```diff
@@ -677,6 +680,14 @@ class Model
         return row_data(row);
     }
 
+    template<typename FilterFn>
+    static std::shared_ptr<FilterModel<ModelData, FilterFn>>
+    filter(std::shared_ptr<Model<ModelData>> source_model, FilterFn filter_fn)
+    {
+        return std::make_shared<FilterModel<ModelData, FilterFn>>(std::move(source_model),
+                                                                  std::move(filter_fn));
+    }
```

Comment:
Sorry, I am missing something. Why is this necessary now?

Could we maybe have real member functions to filter/map/sort the actual model? So that we can write something like:

```c++
auto ui_model = source_model.filter(...).map(...).sort(...);
```

or maybe a something a bit inspired by ranges?

```c++
ui.set_model(source_model | filter(...) | map(...) | sort(...));
```

Reply:
A real member function is not sufficient because there we only have the `this` pointer, but we need a `std::shared_ptr` of the source model. This works in Rust because we implement the `Model` trait for `Rc<T>`.

Reply:
I played a bit with godbold and I think we could have a global `operator|` declared and free standing constructor functions have an API like this:

```cpp
std::shared_ptr<slint::Model<int>> source_model = ...;
auto filtered = source_model | filter_model([](auto n) { return n % 2 == 0; });
```